### ISSUE TITLE
Shard-aware DataLoader streaming for large datasets

### ISSUE THREAD

**Issue by sami-bg** [galilai-group/stable-datasets#54]
`StableDataset` is map-style with an LRU cache of 4 memory-mapped shards, which works well for datasets that fit in 1-2
   shards (~256 MB each), but for ImageNet-1k scale+ datasets (100+ shards) a random `__getitem__` from a shuffled DataLoader would
    thrash the LRU cache because every worker's indices are spread across all shards.

   A good solution would be to add a `to_iterable(seed)` method that returns a `torch.utils.data.IterableDataset` wrapper with shard-aware worker assignment. This would:

   - use `torch.utils.data.get_worker_info()` to partition shards across workers round-robin
   - each worker reads its assigned shards sequentially, and uses within-shard row shuffling
   - this results in peak memory per worker = 1 shard ~= 256MB default
   - `set_epoch(seed)` to re-shuffle shard assignment each epoch
For example:
```python
from stable_datasets.images.imagenet import ImageNet
from torch.utils.data import DataLoader

ds = ImageNet(split="train")  # map-style, shard-backed

for epoch in range(100):
      stream = ds.to_iterable(seed=epoch)  # new IterableDataset wrapper each epoch
      loader = DataLoader(stream, num_workers=4, batch_size=256)

      for batch in loader:
          # Each worker reads only its assigned shards
          # Shard order and within-shard row order are different every epoch
          images, labels = batch["image"], batch["label"]
```
   I propose we keep `StableDataset` itself as map-style so small datasets retain full random access and simpler sampler support.

For testability:

   - Factor shard assignment into a pure function `_assign_shards(num_shards, worker_id, num_workers, seed)` for reproducible unit
   testing
   - Multi-node DDP support (rank-level partitioning) can be added later on top of the same mechanism
   - Handle uneven shard counts (some workers get N+1 shards, others N)
